# Dataset Understanding

## Obesity Risk Intelligence System

This notebook focuses on understanding the dataset used for the Obesity Risk Intelligence System, including its structure, features, target variable, data types, and suitability for the machine learning task.

The findings from this notebook are used to support appropriate preprocessing decisions.

In [15]:
import pandas as pd
import numpy as np

print("Environment is working")

Environment is working


In [16]:
df = pd.read_csv("../data/raw/obesity.csv")

In [17]:
df.head()

,id,Gender,Age,Height,Weight,family_history_with_overweight,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,NObeyesdad
0,0,Male,24.443011,1.699998,81.669950,yes,yes,2.000000,2.983297,Sometimes,no,2.763573,no,0.000000,0.976473,Sometimes,Public_Transportation,Overweight_Level_II
1,1,Female,18.000000,1.560000,57.000000,yes,yes,2.000000,3.000000,Frequently,no,2.000000,no,1.000000,1.000000,no,Automobile,Normal_Weight
2,2,Female,18.000000,1.711460,50.165754,yes,yes,1.880534,1.411685,Sometimes,no,1.910378,no,0.866045,1.673584,no,Public_Transportation,Insufficient_Weight
3,3,Female,20.952737,1.710730,131.274851,yes,yes,3.000000,3.000000,Sometimes,no,1.674061,no,1.467863,0.780199,Sometimes,Public_Transportation,Obesity_Type_III
4,4,Male,31.641081,1.914186,93.798055,yes,yes,2.679664,1.971472,Sometimes,no,1.979848,no,1.967973,0.931721,Sometimes,Public_Transportation,Overweight_Level_II


## 1. Dataset Dimensions

The dataset contains the observations and attributes used to develop the obesity risk classification system.

In [43]:
print("Number of records:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of records: 20758
Number of columns: 18


In [40]:
df.shape

(20758, 18)

The dataset contains 20,758 records and 18 columns. The 18 columns consist of an identifier (`id`), 16 predictive features, and one target variable (`NObeyesdad`).

In [44]:
df.head()

,id,Gender,Age,Height,Weight,family_history_with_overweight,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,NObeyesdad
0,0,Male,24.443011,1.699998,81.669950,yes,yes,2.000000,2.983297,Sometimes,no,2.763573,no,0.000000,0.976473,Sometimes,Public_Transportation,Overweight_Level_II
1,1,Female,18.000000,1.560000,57.000000,yes,yes,2.000000,3.000000,Frequently,no,2.000000,no,1.000000,1.000000,no,Automobile,Normal_Weight
2,2,Female,18.000000,1.711460,50.165754,yes,yes,1.880534,1.411685,Sometimes,no,1.910378,no,0.866045,1.673584,no,Public_Transportation,Insufficient_Weight
3,3,Female,20.952737,1.710730,131.274851,yes,yes,3.000000,3.000000,Sometimes,no,1.674061,no,1.467863,0.780199,Sometimes,Public_Transportation,Obesity_Type_III
4,4,Male,31.641081,1.914186,93.798055,yes,yes,2.679664,1.971472,Sometimes,no,1.979848,no,1.967973,0.931721,Sometimes,Public_Transportation,Overweight_Level_II


In [45]:
df.tail()

,id,Gender,Age,Height,Weight,family_history_with_overweight,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,NObeyesdad
20753,20753,Male,25.137087,1.766626,114.187096,yes,yes,2.919584,3.000000,Sometimes,no,2.151809,no,1.330519,0.196680,Sometimes,Public_Transportation,Obesity_Type_II
20754,20754,Male,18.000000,1.710000,50.000000,no,yes,3.000000,4.000000,Frequently,no,1.000000,no,2.000000,1.000000,Sometimes,Public_Transportation,Insufficient_Weight
20755,20755,Male,20.101026,1.819557,105.580491,yes,yes,2.407817,3.000000,Sometimes,no,2.000000,no,1.158040,1.198439,no,Public_Transportation,Obesity_Type_II
20756,20756,Male,33.852953,1.700000,83.520113,yes,yes,2.671238,1.971472,Sometimes,no,2.144838,no,0.000000,0.973834,no,Automobile,Overweight_Level_II
20757,20757,Male,26.680376,1.816547,118.134898,yes,yes,3.000000,3.000000,Sometimes,no,2.003563,no,0.684487,0.713823,Sometimes,Public_Transportation,Obesity_Type_II


In [41]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20758 entries, 0 to 20757
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              20758 non-null  int64  
 1   Gender                          20758 non-null  str    
 2   Age                             20758 non-null  float64
 3   Height                          20758 non-null  float64
 4   Weight                          20758 non-null  float64
 5   family_history_with_overweight  20758 non-null  str    
 6   FAVC                            20758 non-null  str    
 7   FCVC                            20758 non-null  float64
 8   NCP                             20758 non-null  float64
 9   CAEC                            20758 non-null  str    
 10  SMOKE                           20758 non-null  str    
 11  CH2O                            20758 non-null  float64
 12  SCC                             20758 non-n

In [42]:
df.describe()

,id,Age,Height,Weight,FCVC,NCP,CH2O,FAF,TUE
count,20758.00000,20758.000000,20758.000000,20758.000000,20758.000000,20758.000000,20758.000000,20758.000000,20758.000000
mean,10378.50000,23.841804,1.700245,87.887768,2.445908,2.761332,2.029418,0.981747,0.616756
std,5992.46278,5.688072,0.087312,26.379443,0.533218,0.705375,0.608467,0.838302,0.602113
min,0.00000,14.000000,1.450000,39.000000,1.000000,1.000000,1.000000,0.000000,0.000000
25%,5189.25000,20.000000,1.631856,66.000000,2.000000,3.000000,1.792022,0.008013,0.000000
50%,10378.50000,22.815416,1.700000,84.064875,2.393837,3.000000,2.000000,1.000000,0.573887
75%,15567.75000,26.000000,1.762887,111.600553,3.000000,3.000000,2.549617,1.587406,1.000000
max,20757.00000,61.000000,1.975663,165.057269,3.000000,4.000000,3.000000,3.000000,2.000000


## Dataset Attributes

In [21]:
df.columns.tolist()

['id',
 'Gender',
 'Age',
 'Height',
 'Weight',
 'family_history_with_overweight',
 'FAVC',
 'FCVC',
 'NCP',
 'CAEC',
 'SMOKE',
 'CH2O',
 'SCC',
 'FAF',
 'TUE',
 'CALC',
 'MTRANS',
 'NObeyesdad']

In [22]:
target = "NObeyesdad"

print("Target variable:", target)
print("Number of classes:", df[target].nunique())
print("\nTarget classes:")
print(df[target].unique())

Target variable: NObeyesdad
Number of classes: 7

Target classes:
<ArrowStringArray>
['Overweight_Level_II',       'Normal_Weight', 'Insufficient_Weight',
    'Obesity_Type_III',     'Obesity_Type_II',  'Overweight_Level_I',
      'Obesity_Type_I']
Length: 7, dtype: str


## Target Variable

In [23]:
df[target].value_counts()

NObeyesdad
Obesity_Type_III       4046
Obesity_Type_II        3248
Normal_Weight          3082
Obesity_Type_I         2910
Insufficient_Weight    2523
Overweight_Level_II    2522
Overweight_Level_I     2427
Name: count, dtype: int64

### Target Interpretation

`NObeyesdad` is the target variable that represents the obesity-risk category to be predicted.

This is a supervised multi-class classification problem because the target contains seven different classes.

In [46]:
df["id"].head()

0    0
1    1
2    2
3    3
4    4
Name: id, dtype: int64

### ID Variable

The `id` column is an identifier for individual records. It does not represent a physical, dietary, lifestyle, demographic, or obesity-related characteristic.

Therefore, it should not be used as a predictive feature. Keeping an identifier may allow the model to learn meaningless patterns and can also affect feature interpretation.

The `id` column will therefore be excluded before model training.

## Feature Engineering

In [26]:
numerical_features = [
    "Age",
    "Height",
    "Weight",
    "FCVC",
    "NCP",
    "CH2O",
    "FAF",
    "TUE"
]

binary_features = [
    "Gender",
    "family_history_with_overweight",
    "FAVC",
    "SMOKE",
    "SCC"
]

ordinal_features = [
    "CAEC",
    "CALC"
]

nominal_features = [
    "MTRANS"
]

In [27]:
print("Numerical features:", len(numerical_features))
print("Binary categorical features:", len(binary_features))
print("Ordinal features:", len(ordinal_features))
print("Nominal features:", len(nominal_features))

print("\nTotal predictive features:",
      len(numerical_features)
      + len(binary_features)
      + len(ordinal_features)
      + len(nominal_features))

Numerical features: 8
Binary categorical features: 5
Ordinal features: 2
Nominal features: 1

Total predictive features: 16


In [28]:
df.dtypes

id                                  int64
Gender                                str
Age                               float64
Height                            float64
Weight                            float64
family_history_with_overweight        str
FAVC                                  str
FCVC                              float64
NCP                               float64
CAEC                                  str
SMOKE                                 str
CH2O                              float64
SCC                                   str
FAF                               float64
TUE                               float64
CALC                                  str
MTRANS                                str
NObeyesdad                            str
dtype: object

In [29]:
for feature in numerical_features:
    print(feature, "->", df[feature].dtype)

Age -> float64
Height -> float64
Weight -> float64
FCVC -> float64
NCP -> float64
CH2O -> float64
FAF -> float64
TUE -> float64


In [30]:
for feature in binary_features + ordinal_features + nominal_features:
    print(feature, "->", df[feature].dtype)

Gender -> str
family_history_with_overweight -> str
FAVC -> str
SMOKE -> str
SCC -> str
CAEC -> str
CALC -> str
MTRANS -> str


In [36]:
categorical_features = (
    binary_features +
    ordinal_features +
    nominal_features
)

for feature in categorical_features:
    print(f"\n--- {feature} ---")
    print(df[feature].value_counts())


--- Gender ---
Gender
Female    10422
Male      10336
Name: count, dtype: int64

--- family_history_with_overweight ---
family_history_with_overweight
yes    17014
no      3744
Name: count, dtype: int64

--- FAVC ---
FAVC
yes    18982
no      1776
Name: count, dtype: int64

--- SMOKE ---
SMOKE
no     20513
yes      245
Name: count, dtype: int64

--- SCC ---
SCC
no     20071
yes      687
Name: count, dtype: int64

--- CAEC ---
CAEC
Sometimes     17529
Frequently     2472
Always          478
no              279
Name: count, dtype: int64

--- CALC ---
CALC
Sometimes     15066
no             5163
Frequently      529
Name: count, dtype: int64

--- MTRANS ---
MTRANS
Public_Transportation    16687
Automobile                3534
Walking                    467
Motorbike                   38
Bike                        32
Name: count, dtype: int64


In [33]:
print(df["CAEC"].value_counts())

CAEC
Sometimes     17529
Frequently     2472
Always          478
no              279
Name: count, dtype: int64


In [34]:
print(df["CALC"].value_counts())

CALC
Sometimes     15066
no             5163
Frequently      529
Name: count, dtype: int64


In [35]:
print(df["CAEC"].unique())
print(df["CALC"].unique())

<ArrowStringArray>
['Sometimes', 'Frequently', 'no', 'Always']
Length: 4, dtype: str
<ArrowStringArray>
['Sometimes', 'no', 'Frequently']
Length: 3, dtype: str


## Preprocessing Implications of Feature Types

The feature types identified above are important because different types of variables require different preprocessing strategies.

### Numerical Features

Age, Height, Weight, FCVC, NCP, CH2O, FAF, and TUE are numerical features. Their numerical ranges should be considered during preprocessing, and appropriate numerical transformations such as scaling may be applied.

### Binary Categorical Features

Gender, family_history_with_overweight, FAVC, SMOKE, and SCC are categorical features with two categories. They require categorical encoding before being provided to machine-learning algorithms.

### Ordinal Features

CAEC and CALC contain categories with a meaningful order. Their preprocessing should preserve this order.

### Nominal Features

MTRANS contains categories without a natural ordering. Its encoding should therefore avoid introducing an artificial ranking between categories.

In [37]:
df[numerical_features].describe().T

,count,mean,std,min,25%,50%,75%,max
Age,20758.0,23.841804,5.688072,14.00,20.000000,22.815416,26.000000,61.000000
Height,20758.0,1.700245,0.087312,1.45,1.631856,1.700000,1.762887,1.975663
Weight,20758.0,87.887768,26.379443,39.00,66.000000,84.064875,111.600553,165.057269
FCVC,20758.0,2.445908,0.533218,1.00,2.000000,2.393837,3.000000,3.000000
NCP,20758.0,2.761332,0.705375,1.00,3.000000,3.000000,3.000000,4.000000
CH2O,20758.0,2.029418,0.608467,1.00,1.792022,2.000000,2.549617,3.000000
FAF,20758.0,0.981747,0.838302,0.00,0.008013,1.000000,1.587406,3.000000
TUE,20758.0,0.616756,0.602113,0.00,0.000000,0.573887,1.000000,2.000000


### Numerical Feature Ranges

The numerical features have different ranges and statistical properties. This needs to be considered when preparing the data for machine learning, particularly for algorithms that can be affected by differences in feature scale.

## Preprocessing Handoff

Based on the dataset understanding and feature classification, the following preprocessing requirements were identified:

| Feature Group | Features | Preprocessing Consideration |
|---|---|---|
| Numerical | Age, Height, Weight, FCVC, NCP, CH2O, FAF, TUE | Numerical preprocessing/scaling |
| Binary categorical | Gender, family_history_with_overweight, FAVC, SMOKE, SCC | Categorical encoding |
| Ordinal | CAEC, CALC | Encoding that preserves category order |
| Nominal | MTRANS | Encoding without imposing artificial order |
| Identifier | id | Remove before model training |
| Target | NObeyesdad | Keep separate from input features |
